### DISCUSSION TITLE
TestStore type inference not working

### DISCUSSION THREAD

**Discussion by lukeredpath** [pointfreeco/swift-composable-architecture#1028] [General]
Since updating to Xcode 13.3 the compiler no longer seems to be able to infer the type of our `TestStore` in any of our main app target tests, which makes it very hard to use without any auto-completion of actions in send and receive calls or state properties inside assertion closures.

Can anyone think of what might be breaking this? I spun up a new project and added TCA and an example test and it infers the `TestStore` type just fine but I don't know what it could be about our project causing this to fail.

For example:

```swift
let store = TestStore(
    initialState: SessionState.loggedInAsBot,
    reducer: sessionReducer,
    environment: SessionEnvironment.failing
)
```

I'm being as explicit as I can here. This is taken from one of our tests. In this example:

* `SessionState` is a static var of type `SessionState`.
* `SessionEnvironment.failing` is a static var of type `SessionEnvironment`
* If I option-click on `sessionReducer` it correctly shows me the type `Reducer<SessionState, SessionAction, SessionEnvironment>`

Yet when I option-click on `store` it gives me a type definition of `let store: <<error type>>` which means it cannot infer the type.

This only seems to be a problem in our main app target. All of the feature tests in our SPM feature module test targets are working as expected.

**Comment by jshier** [comment]
If Xcode couldn't infer types properly you'd be seeing a build failure, not just a failure in Xcode's features. Are you seeing a build failure? If not, I'm surprised you've never seen this before, as Xcode's autocomplete often fails, especially in test targets. Narrowing down the cause and reporting bugs is the only way to fix it.

**Reply by lukeredpath** [reply]
No, it compiles fine. We've come across occasions where autocomplete doesn't work properly but never so consistently on such a fundamental type in our entire app test target. I'm wondering now if this might be one of those weird bugs that goes away if I blow away my derived data. I'm going to try on another laptop too.

One place we've seen auto-complete fail completely is in TCA reducers as soon as we chain on any high-level reducer. For this reason we've had to adopt a convention of defining our main reducer as `_reducer` and then declaring a separate `let reducer = _reducer.chain().stuff().here()`.

**Comment by lukeredpath** [comment]
I've tried blowing away my entire derived data folder and letting Xcode re-index but still no luck.

Strangely, changing to this:

```swift
let store = TestStore<SessionState, SessionState, SessionAction, SessionAction, SessionEnvironment>(...)
```

Makes no difference. But this:

```swift
let store: TestStore<SessionState, SessionState, SessionAction, SessionAction, SessionEnvironment> = .init(...)
```

Does work and I'm able to auto-complete again. Its not very ergonomic though.

**Reply by jshier** [reply]
Strange. What if you use the `.init` on the type directly?
